### PR TITLE
Optimize RAFS filesystem builder

### PR DESCRIPTION
**Pull request by jiangliu**
rafs: optimize the way to build RAFS filesystem
    
The current way to build RAFS filesystem is:
- build the lower tree from parent bootstrap
- convert the lower tree into an array
- build the upper tree from source
- merge the upper tree into the lower tree
- convert the merged tree into another array
- dump nodes from the array
    
Now we optimize it as:
- build the lower tree from parent bootstrap
- build the upper tree from source
- merge the upper tree into the lower tree
- dump the merged tree

### CODE REVIEW

File: rafs/src/builder/core/v5.rs [hunk 1]
```diff
@@ -187,7 +173,7 @@ impl Bootstrap {
 
         // Set super block
         let mut super_block = RafsV5SuperBlock::new();
-        let inodes_count = bootstrap_ctx.inode_map.len() as u64;
+        let inodes_count = bootstrap_ctx.inode_map.len() as u64 + 1;
```
Comment (bergwolf) [inline rafs/src/builder/core/v5.rs:176]: Can you separate this commit into a different PR, and add a full analysis of it? It is quite easy to get slipped in w/o notice but we definitely need to figure out what happened and get it backported.

Comment (jiangliu) [reply rafs/src/builder/core/v5.rs]: A dedicated PR for it: https://github.com/dragonflyoss/image-service/pull/1275

**Review by imeoer — COMMENTED** [review]
Thanks for the hard work! Do we have a perf comparison? For example, convert a wordpress image.

File: rafs/src/builder/core/context.rs [hunk 2]
```diff
@@ -962,14 +964,14 @@ pub struct BootstrapContext {
     /// This build has a parent bootstrap.
     pub layered: bool,
     /// Cache node index for hardlinks, HashMap<(layer_index, real_inode, dev), Vec<index>>.
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply rafs/src/builder/core/context.rs:966]: fixed

File: rafs/src/builder/core/tree.rs [hunk 3]
```diff
@@ -55,204 +61,188 @@ impl Tree {
...
-                    return self.apply(target, false, whiteout_spec);
-                }
-                return self.remove(target, whiteout_type, origin_name, parent_name);
-            }
-        }
-
-        let target_paths = target.target_vec();
-        let target_paths_len = target_paths.len();
-        let depth = self.node.target_vec().len();
+    /// Walk all nodes in pre DFS mode.
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply rafs/src/builder/core/tree.rs:115]: fixed

**Comment by jiangliu** [comment]
> Thanks for the hard work! Do we have a perf comparison? For example, convert a wordpress image.

There is another PR to improve the performance, will do test with the coming PR:)

**Review by imeoer — APPROVED** [review]
